content/development/testing.md

```diff
@@ -9,7 +9,316 @@ chapter: false
+                proxy_set_header Host $host;
+                proxy_set_header Proxy "";
+                proxy_set_header Upgrade $http_upgrade;
+                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+                proxy_set_header X-Forwarded-Port $server_port;
+                proxy_set_header X-Forwarded-Proto $scheme;
+
+                proxy_http_version 1.1;
+                proxy_buffering off;
+                proxy_connect_timeout 60s;
+                proxy_read_timeout 36000s;
+                proxy_redirect off;
+
+                proxy_pass_header Authorization;
+        }
+```
+
+In both cases (Apache2, Nginx) you have to change your `modsecurity.conf` settings. Open that file and find the directive `SecResponseBodyMimeType`. Modify the arguments:
+
+```
+SecResponseBodyMimeType text/plain text/html text/xml application/json
+```
+
+Note, that the default value does not have the mime type `application/json`.
+
+In your `crs-setup.conf` you need to add these extra rules (after the rule `900990`):
+
+```
+SecAction
+    "id:900005,\
+    phase:1,\
+    nolog,\
+    pass,\
+    ctl:ruleEngine=DetectionOnly,\
+    ctl:ruleRemoveById=910000,\
+    setvar:tx.blocking_paranoia_level=4,\
+    setvar:tx.crs_validate_utf8_encoding=1,\
+    setvar:tx.arg_name_length=100,\
+    setvar:tx.arg_length=400,\
+    setvar:tx.max_file_size=64100,\
+    setvar:tx.combined_file_sizes=65535"
+
+SecRule REQUEST_HEADERS:X-CRS-Test "@rx ^.*$" \
+    "id:999999,\
+    phase:1,\
+    pass,\
+    t:none,\
+    log,\
+    msg:'%{MATCHED_VAR}'"
+```
+
+Now after the restart all request will send to `httpbin`. Let's start testing.
+
+## Go-ftw
+
+Tests are performed using a golang tool called [go-ftw](https://github.com/coreruleset/go-ftw). We run them using a [GitHub actions pipeline](https://github.com/coreruleset/coreruleset/blob/{{< param crs_dev_branch >}}/.github/workflows/test.yml). You can easily reproduce that locally, in your workstation.
+
+For that you will need:
+
+- the coreruleset git repository
+- docker and docker-compose (modern versions of docker already include compose functionality)
+  OR
+  an own instance (see above installation steps)
+- golang compiler
+- your rules and tests!
+
+You can download the pre-compiled version or build your own instance. Precompiled versions are [here](https://github.com/coreruleset/go-ftw/releases). Just download it and you can use that.
+
+If you want to build one, you can type:
+
+```bash
+$ go install github.com/coreruleset/go-ftw@latest
+```
+
+This will build the binary into your `$HOME/go/bin` directory. Other option:
+
+```bash
+$ git clone https://github.com/coreruleset/go-ftw.git
+$ cd go-ftw
+$ go build .
+```
+
+This will build the binary into your source directory.
+
+Now create a configuration file. Because the two webservers uses different log files, and - perhaps - different ports, you can create two different config files.
```

Comment:
```suggestion
Now create a configuration file. Because Apache httpd and nginx use different log file paths, and, perhaps, different ports, you may want to create two different configuration files for **go-ftw**.
```